katsdpcontroller/product_config.py

```diff
@@ -1,22 +1,73 @@
 """Support for manipulating product config dictionaries."""
 
+import collections.abc
 import logging
 import itertools
-import json
-import urllib
 import copy
+import math
 from abc import ABC, abstractmethod
 from distutils.version import StrictVersion
+from typing import (
+    Tuple, List, Dict, Mapping, AbstractSet, Sequence, Iterable,
+    Union, ClassVar, Type, TypeVar, Optional, Any, cast, TYPE_CHECKING
+)
 
-import jsonschema
+import networkx
+import yarl
 
+import katpoint
 from katsdptelstate.endpoint import endpoint_list_parser
 import katportalclient
 
 from . import schemas
 
 
 logger = logging.getLogger(__name__)
+_S = TypeVar('_S', bound='Stream')
+_ValidTypes = Union[AbstractSet[str], Sequence[str]]
+#: Minimum observation time for continuum imager (seconds)
+DEFAULT_CONTINUUM_MIN_TIME = 15 * 60.0     # 15 minutes
```

Comment:
I don't mind making a `defaults.py` to hold things like this. It's not totally clear to me which things should move there though. My candidates would be:

```python
INGEST_GPU_NAME = 'GeForce GTX TITAN X'
#: Maximum number of custom signals requested by (correlator) timeplot
TIMEPLOT_MAX_CUSTOM_SIGNALS = 256
#: Target size of objects in the object store
WRITER_OBJECT_SIZE = 20e6    # 20 MB
#: Maximum channels per chunk for spectral imager
SPECTRAL_OBJECT_CHANNELS = 128
#: Minimum observation time for continuum imager (seconds)
DEFAULT_CONTINUUM_MIN_TIME = 15 * 60.0     # 15 minutes
#: Minimum observation time for spectral imager (seconds)
DEFAULT_SPECTRAL_MIN_TIME = 3600.0         # 1 hour
#: Size of cal buffer in seconds
DEFAULT_CAL_BUFFER_TIME = 25 * 60.0        # 25 minutes (allows a single batch of 15 minutes)
#: Maximum number of scans to include in report
DEFAULT_CAL_MAX_SCANS = 1000
#: Speed at which flags are transmitted, relative to real time
DEFAULT_FLAGS_RATE_RATIO = 8.0
#: Alignment constraint for `int_time` in katcbfsim
KATCBFSIM_SPECTRA_PER_HEAP = 256
```

Can you look though generator.py and product_config.py and let me know if there are others you'd want to move?

Reply:
Looks like a good list - I may be tempted to put the _VOL definitions in there as well, unless there is a reason to keep them in generator.

Reply:
I consider the _VOL constants to be closely tied to the details of how containers are run, rather than being tunable knobs like the others. I'd also rather not have a dependency on `scheduler`.

Reply:
I've moved the items from that list into a defaults.py. I'm not totally convinced that KATCBFSIM_SPECTRA_PER_HEAP truly belongs there, because it's describing a property of an external system rather than being an independently twiddlable knob, but that's getting into bikeshed territory.